Generate Recovery Codes

J0 opened the pull request:
## What kind of change does this PR introduce?

This PR  should add the ability for a user signed in with a single factor(or more factors) to call `/mfa/%s/generate_recovery_codes` to produce recovery codes which will eventually be used to `SignIn`, recover a factor. 

Given the all encompassing nature of this endpoint it should eventually move out of the `/mfa` route.

There should also be a rate limiter on this but haven't decided on value.

## What is the current behavior?
None

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

J0 commented on api/admin.go:
```diff
@@ -175,7 +175,7 @@ func (a *API) adminUserUpdate(w http.ResponseWriter, r *http.Request) error {
 		}
 
-		if terr := models.NewAuditLogEntry(tx, instanceID, adminUser, models.UserModifiedAction, "",  map[string]interface{}{
+		if terr := models.NewAuditLogEntry(tx, instanceID, adminUser, models.UserModifiedAction, "", map[string]interface{}{
```
Gofmt change

J0 replied:
from Gofmt -s -w.

J0 commented on models/audit_log_entry.go:
```diff
@@ -48,15 +49,16 @@ var actionLogTypeMap = map[AuditAction]auditLogType{
-	Payload JSONMap `json:"payload" db:"payload"`
-	CreatedAt time.Time `json:"created_at" db:"created_at"`
-	IPAddress string `json:"ip_address" db:"ip_address"`
+	Payload    JSONMap   `json:"payload" db:"payload"`
```
from `gofmt -s -w .`

J0 commented on api/mfa.go:
```diff
@@ -51,3 +58,45 @@ func (a *API) DisableMFA(w http.ResponseWriter, r *http.Request) error {
+		if terr != nil {
+			return internalServerError("Error creating recovery code").WithInternalError(terr)
+		}
+		recoveryCodes = append(recoveryCodes, recoveryCode)
```
This portion is a little iffy, it assumes that the DB write went through -- not sure if it'd make sense to read from db again and return that or return what was generated assuming the DB write was ok

kangmingtay replied:
yeah good point, i think its safe to make those assumptions as long as:
1. we check if the transaction below errors out 
2. if the transaction completes successfully, we can be sure that all the writes went through

kangmingtay commented on models/recovery_code.go:
```diff
@@ -0,0 +1,51 @@
+		ID:           id,
+		UserID:       user.ID,
+		RecoveryCode: recoveryCode,
+		CreatedAt:    now,
```
gobuffalo will automatically set this for you (https://gobuffalo.io/documentation/database/models/#unix-timestamps)

J0 replied:
Ah yup this is intentionally done to skirt Buffalo's default behaviour so that each batch of recovery codes has the same timestamp and users can query batches of codes by timestamp. Otherwise there would be slight difference between the timestamp recorded but it's a good point -- I'm not sure if :

1. The time difference is significant since it's probably on the order of ms or lower
2. It might be better to have an explicit timestamp field

Let me check what the timestamps in DB are and adjust accordingly

J0 replied:
Yeah reverted so we use the default created by Buffalo -- users can check the audit log to tally timestamps if they want to look into when a batch was created

kangmingtay commented on models/recovery_code.go:
```diff
@@ -0,0 +1,51 @@
+// FindValidRecoveryCodes returns all valid recovery codes associated to a user
+func FindValidRecoveryCodesByUser(tx *storage.Connection, user *User) ([]*RecoveryCode, error) {
+	recoveryCodes := []*RecoveryCode{}
+	if err := tx.Q().Where("user_id = ? AND valid = ?", user.ID, true).All(&recoveryCodes); err != nil {
```
hmm i don't see a `valid` field in the `mfa_recovery_codes` table?

J0 replied:
You are right, my bad on this one. Made a last minute edit  and didn't update -- should add a test for this as well.

kangmingtay commented on crypto/crypto.go:
```diff
@@ -13,8 +13,12 @@ import (
 // SecureToken creates a new random token
-func SecureToken() string {
-	b := make([]byte, 16)
+func SecureToken(options ...int) string {
```
why do we take in a variable argument list here?

J0 replied:
Oh this is a way of [taking in optional arguments](https://stackoverflow.com/questions/2032149/optional-parameters-in-go). `SecureToken()` is used to generate recovery codes which are of length 8 instead of the default of length 16, passing in an int (e.g. SecureToken(8)) will return a  length 8 random string

kangmingtay commented on api/mfa.go:
```diff
@@ -51,3 +58,45 @@ func (a *API) DisableMFA(w http.ResponseWriter, r *http.Request) error {
+		}
+		return nil
+	})
+
```
we need to check for the err here in case the transaction fails (e.g. one of the inserts for the recovery codes fail) and return a 500

kangmingtay commented on api/mfa.go:
```diff
@@ -51,3 +58,45 @@ func (a *API) DisableMFA(w http.ResponseWriter, r *http.Request) error {
 }
+
+func (a *API) GenerateRecoveryCodes(w http.ResponseWriter, r *http.Request) error {
+	const NUM_RECOVERY_CODES = 8
+	const RECOVERY_CODE_LENGTH = 8
```
nit: as much as i hate to say this, the standard convention in go is to use camelCase and the first letter is either an uppercase or lowercase depending on whether you want it to be exported or not.

J0 commented:
Going to merge first to aid development of other branches, have addressed the comments but feel free to let me know if there are additional changes we should make -- will file a PR against the `mfa` branch
